### PR TITLE
feat: support segment midpoints in line editor

### PR DESCRIPTION
**Pull request by ad1992**
https://user-images.githubusercontent.com/11256141/187417807-3efeb673-6c96-4744-be0e-70119b0c6839.mp4

TODOS

- [x] Midpoint inside editor for line segments with 0 roughness and sharp edges
- [x] Midpoint inside editor for line segments with 0 roughness and round edges
- [x] Midpoint inside editor for line segments with non zero roughness and round edges
- [x] Don't show midpoints for short segments with sharp edges
- [x]  Don't show midpoints for short segments with round edges
- [x] The current mid point algo for beizer is not very accurate so improve it
- [x] cache editor midpoints for better perf
- [x] fix undo/redo

### CODE REVIEW

**Comment by dwelle** [comment]
I think the hiding based on proximity behavior from the other PR should apply here as well. Is the logic reusable and thus introduce it after we merge the other PR, or do we need to implement separately?

**Comment by ad1992** [comment]
> I think the hiding based on proximity behavior from the other PR should apply here as well. Is the logic reusable and thus merge this now and introduce the point hiding after we merge the other PR, or do we need to implement separately?

It can be reused but I think inside editor its ok to show all the points, as thats what we are doing even for 2 point arrows (inside editor we are not hiding in that PR) as inside editor user should be able to edit all points.
Later if we want we can hide inside editor.
As per this PR, what I can try is not show mid points unless segment length is > x ?

**Comment by dwelle** [comment]
Oh, I didn't mean existing points, but the phantom points. If there's too little space between two existing points, let's not suggest a phantom point as it unlikely you'll want to create one in there anyway.

**Comment by ad1992** [comment]
> Oh, I didn't mean existing points, but the phantom points. If there's too little space between two existing points, let's not suggest a phantom point as it unlikely you'll want to create one in there anyway.

yep thats what I suggested above too, lets not show mid point in segment for smaller segments

**Comment by dwelle** [comment]
> > Oh, I didn't mean existing points, but the phantom points. If there's too little space between two existing points, let's not suggest a phantom point as it unlikely you'll want to create one in there anyway.
> 
> yep thats what I suggested above too, lets not show mid point in segment for smaller segments

Missed that. Good, let's do it. 👍

File: src/math.ts [hunk 1]
```diff
@@ -348,3 +349,74 @@ export const getPointsInBezierCurve = (
...
+export const getBezierCurveLength = (
+  element: NonDeleted<ExcalidrawLinearElement>,
+  endPoint: Point,
+) => {
+  const arcLengths = getBezierCurveArcLengths(element, endPoint);
+  return arcLengths.at(-1) as number;
+};
+
+// This maps u to actual t on the curve so that when t = 0.5, its actually the point at 50% of the length
+export const mapUToBezierT = (
```
Comment (dwelle) [inline src/math.ts:387]: we should clarify what `u`/`t` stand for, or better yet use a different name for the function

File: src/element/linearElementEditor.ts [hunk 2]
```diff
@@ -359,15 +363,15 @@ export class LinearElementEditor {
     };
   }
 
-  static isHittingMidPoint = (
+  static getSegmentMidpointHitCoords = (
```
Comment (dwelle) [inline src/element/linearElementEditor.ts:366]: this function must be optimized as it can take 150+ ms per frame for 50-point line (common in libraries) when called in the pointermove handler to determine which (mid)point you're hovering over — which causes noticeable lag, even race conditions where points tend to stay in hover state even after you move the cursor away.

If we can't make the algo better, e.g. narrow down to the segment the cursor is in instead of computing cursor distance to every line segment center, then we'll need to precalculate all segment midpoints and cache on line editor construction (it would also need to update on point move/delete/add/etc.).

**Comment by dwelle** [comment]
Works great ❤️.

Haven't fully reviewed the math, which may or may not be connected to the perf issue pointed above. Solution is either use better math / heuristic to not compute as many points as you're doing right now, or caching.

**Comment by dwelle** [comment]
- [ ] In specific cases it creates the new point in a wrong segment (not related to caching). From what I gathered the conditions are 1) at least one phantom mid-point is hidden due to proximity 2) there's at last one more segment after the segment in which you're adding the mid-point

  excal-midpoints-wrong-new-point
  
  ```
  {"type":"excalidraw/clipboard","elements":[{"type":"line","version":761,"versionNonce":84824830,"isDeleted":false,"id":"DuAsvTnKAEwdo-LDPKnhG","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":635,"y":221,"strokeColor":"#000000","backgroundColor":"transparent","width":226,"height":156,"seed":2074270108,"groupIds":[],"strokeSharpness":"sharp","boundElements":[],"updated":1662992934399,"link":null,"locked":false,"startBinding":null,"endBinding":null,"lastCommittedPoint":null,"startArrowhead":null,"endArrowhead":null,"points":[[0,0],[156,-86],[176,-77],[226,21],[66,70]]},{"type":"rectangle","version":30,"versionNonce":1475535842,"isDeleted":false,"id":"51WkiFZ4vNg5vQnWvfDk1","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":786,"y":173,"strokeColor":"#c92a2a","backgroundColor":"transparent","width":120,"height":42,"seed":1518143535,"groupIds":[],"strokeSharpness":"sharp","boundElements":[],"updated":1662992934399,"link":null,"locked":false}],"files":{}}
  ```

- midpoints position not being updated in many cases. Some I found:
  - [ ] during collaboration when someone moves your line while you're in the editor 
  - [ ] you update settings that change the line's shape (e.g. edge sharpness or sloppiness)
  - [ ] undo/redo: e.g. you create a mid-point and undo it, or you move a point and undo the change
  - [ ] duplicating a point: 1) select a point 2) cmd-D (or use the duplicate button in sidebar)

**Comment by ad1992** [comment]
Pushed the fixes including the one for collab flickering, should work fine now

**Comment by dwelle** [comment]
works sweet ❤️

File: src/components/App.tsx [hunk 3]
```diff
@@ -2718,15 +2720,19 @@ class App extends React.Component<AppProps, AppState> {
         event,
         scenePointerX,
         scenePointerY,
-        this.state.editingLinearElement,
-        this.state.gridSize,
+        this.state,
       );
+      if (!editingLinearElement) {
+        return;
+      }
```
Comment (dwelle) [inline src/components/App.tsx:2727]: this is annoying since it should never happen in practice (and if it does I'm not sure exiting is the right thing to do here), but it requires you to read the LinearEditor code to find out. IOW, it's not semantic.

Solutions are: 

- make `handlePointerMove`  appState param into `MarkNonNullable<AppState, "editingLinearElement">`, but it'd require either asserting in callsites, or using a type guard — which may not be that terrible: `if(isPropNonNullable(appState, "editingLinearElement"))`
- pass `editingLinearElement` as another arg
- do nothing...

File: src/math.ts [hunk 4]
```diff
@@ -348,7 +347,7 @@ export const getPointsInBezierCurve = (
     t -= 0.05;
   }
   if (pointsOnCurve.length) {
-    if (!LinearElementEditor.isEqual(pointsOnCurve.at(-1)!, endPoint)) {
+    if (arePointsEqual(pointsOnCurve[1], endPoint)) {
```
Comment (dwelle) [inline src/math.ts:350]: why this change?

File: src/element/linearElementEditor.ts [hunk 5]
```diff
@@ -72,15 +84,15 @@ export class LinearElementEditor {
       lastClickedPoint: -1,
     };
     this.hoverPointIndex = -1;
-    this.midPointHovered = false;
+    this.segmentMidPointHoveredCoords = null;
+    LinearElementEditor.updateEditorMidPointsCache(element, appState);
```
Comment (dwelle) [inline src/element/linearElementEditor.ts:88]: unless I'm missing something, updating the cache anywhere is redundant because we already 
do it inside the getter, no?

File: src/math.ts (hunk 4 above)
Comment (ad1992) [reply src/math.ts]: Reverted. I might remove this as well later if the midpoints calculated fine without the last point but for now I will keep it.

File: src/element/linearElementEditor.ts (hunk 5 above)
Comment (ad1992) [reply src/element/linearElementEditor.ts:88]: You are right, we don't need it anymore as now its updated on version change.

File: src/renderer/renderScene.ts [hunk 6]
```diff
@@ -222,7 +222,7 @@ const renderLinearPointHandles = (
     appState,
   ).filter((midPoint) => midPoint !== null) as Point[];
 
-  if (!midPoints || !midPoints?.length) {
+  if (appState.editingLinearElement && !midPoints.length) {
     return;
   }
```
Comment (dwelle) [inline src/renderer/renderScene.ts:227]: we shouldn't return here without restoring the context (which I assumed caused the zoom bug). There shouldn't be much penalty for looping through an empty array (and it should happen in common case anyway?).

**Review by dwelle — APPROVED** [review]
